src/Turtle/Bytes/Lazy.hs

```diff
@@ -0,0 +1,422 @@
+ stderr as a lazy `ByteString`
+
+    This command is more powerful than `proc`, but highly vulnerable to code
+    injection if you template the command line with untrusted input
+-}
+shellStreamWithErr
+    :: MonadIO io
+    => Text
+    -- ^ Command line
+    -> Shell ByteString
+    -- ^ Chunks of bytes written to process input
+    -> io (ExitCode, ByteString, ByteString)
+    -- ^ (Exit code, stdout, stderr)
+shellStreamWithErr cmdline =
+    streamWithErr (Process.shell (Data.Text.unpack cmdline))
+
+-- | Halt an `Async` thread, re-raising any exceptions it might have thrown
+halt :: Async a -> IO ()
+halt a = do
+    m <- Async.poll a
+    case m of
+        Nothing        -> Async.cancel a
+        Just (Left  e) -> Exception.throwIO e
+        Just (Right _) -> return ()
+
+{-| `system` generalizes `shell` and `proc` by allowing you to supply your own
+    custom `CreateProcess`.  This is for advanced users who feel comfortable
+    using the lower-level @process@ API
+-}
+system
+    :: MonadIO io
+    => Process.CreateProcess
+    -- ^ Command
+    -> Shell ByteString
+    -- ^ Chunks of bytes written to process input
+    -> io ExitCode
+    -- ^ Exit code
+system p s = liftIO (do
+    let open = do
+            (m, Nothing, Nothing, ph) <- Process.createProcess p
+            case m of
+                Just hIn -> System.IO.hSetBuffering hIn (System.IO.BlockBuffering Nothing)
+                _        -> return ()
+            return (m, ph)
+
+    -- Prevent double close
+    mvar <- MVar.newMVar False
+    let close handle = do
+            MVar.modifyMVar_ mvar (\finalized -> do
+                Control.Monad.unless finalized
+                    (ignoreSIGPIPE (System.IO.hClose handle))
+                return True )
+    let close' (Just hIn, ph) = do
+            close hIn
+            Process.terminateProcess ph
+        close' (Nothing , ph) = do
+            Process.terminateProcess ph
+
+    let handle (Just hIn, ph) = do
+            let feedIn :: (forall a. IO a -> IO a) -> IO ()
+                feedIn restore =
+                    restore (ignoreSIGPIPE (outhandle hIn s))
+                    `Exception.finally` close hIn
+            Exception.mask (\restore ->
+                Async.withAsync (feedIn restore) (\a ->
+                    restore (Process.waitForProcess ph) <* halt a ) )
+        handle (Nothing , ph) = do
+            Process.waitForProcess ph
+
+    Exception.bracket open close' handle )
+
+{-| `stream` generalizes `inproc` and `inshell` by allowing you to supply your
+    own custom `CreateProcess`.  This is for advanced users who feel comfortable
+    using the lower-level @process@ API
+
+    Throws an `ExitCode` exception if the command returns a non-zero exit code
+-}
+stream
+    :: MonadIO io
+    => Process.CreateProcess
+    -- ^ Command
+    -> Shell ByteString
+    -- ^ Chunks of bytes written to process input
+    -> io (ExitCode, ByteString)
```

Comment:
Actually, something seems wrong here.  This type implies that we have access to the `ExitCode` as the same time as the `ByteString`, which is being read lazily.  If we are able to access the `ExitCode`, wouldn't that imply that the reading `ByteString` would fail due to the subprocess handle being closed?

Reply:
I guess that I should have port the initial stream version as well then... I should really try and test these functions with some examples, I'm not sure whether the out handle will be accessible or not in that case.

Reply:
One option would be to read the lazy `ByteString` strictly (i.e. force the entire `ByteString`) so that you don't have to worry about still reading it after the handle is closed